commonmark-ext-metadata/src/main/java/org/commonmark/ext/metadata/internal/MetadataBlockParser.java

```diff
@@ -0,0 +1,121 @@
+package org.commonmark.ext.metadata.internal;
+
+import org.commonmark.ext.metadata.MetadataBlock;
+import org.commonmark.ext.metadata.MetadataNode;
+import org.commonmark.internal.DocumentBlockParser;
+import org.commonmark.node.Block;
+import org.commonmark.parser.InlineParser;
+import org.commonmark.parser.block.*;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class MetadataBlockParser extends AbstractBlockParser {
+    private static final Pattern REGEX_METADATA = Pattern.compile("^[ ]{0,3}([A-Za-z0-9_-]+):\\s*(.*)");
+    private static final Pattern REGEX_METADATA_LIST = Pattern.compile("^[ ]+-\\s*(.*)");
+    private static final Pattern REGEX_METADATA_LITERAL = Pattern.compile("^\\s*(.*)");
+    private static final Pattern REGEX_BEGIN = Pattern.compile("^-{3}(\\s.*)?");
+    private static final Pattern REGEX_END = Pattern.compile("^(-{3}|\\.{3})(\\s.*)?");
+
+    private List<String> lines;
+    private MetadataBlock block;
+
+    public MetadataBlockParser() {
+        lines = new ArrayList<>();
+        block = new MetadataBlock();
+    }
+
+    @Override
+    public Block getBlock() {
+        return block;
+    }
+
+    @Override
+    public void addLine(CharSequence line) {
+        lines.add(line.toString());
+    }
+
+    @Override
+    public BlockContinue tryContinue(ParserState parserState) {
+        final CharSequence line = parserState.getLine();
+
+        if (REGEX_END.matcher(line).matches()) {
+            // if this line is `---` or `...` which means end of metadata block
+            return BlockContinue.finished();
+        } else if (!REGEX_METADATA.matcher(line).matches() && !REGEX_METADATA_LIST.matcher(line).matches() &&
+                !REGEX_METADATA_LITERAL.matcher(line).matches()) {
```

Comment:
This regex is `^\\s*(.*)`, won't this just always match? Neither space nor any character is required.

What I meant with doing the matching is moving the logic from `parseInlines` to `tryContinue`. Instead of iterating over the lines in `parseInlines`, keep the state in the class and process one line at a time. If a line doesn't match anything, finish the block.

I can take a stab at it if you want.